panic in portallocator when updating service

Hi, All,

When  I was updating service, panic happened in docker engine's log, like below:
```
time="2016-09-01T12:03:35.697250351+08:00" level=error msg="Handler for GET /v1.24/services/dy7yajeu246ztckluq367wvfo returned error: service dy7yajeu246ztckluq367wvfo not found" 
time="2016-09-01T12:03:55.532280461+08:00" level=error msg="Error updating service adnfj9n8um05: rpc error: code = 5 desc = service adnfj9n8um05 not found" 
time="2016-09-01T12:03:55.532477499+08:00" level=error msg="Handler for POST /v1.24/services/adnfj9n8um05/update returned error: rpc error: code = 5 desc = service adnfj9n8um05 not found" 
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x10 pc=0x14b2dcc]

goroutine 264 [running]:
panic(0x1a619a0, 0xc820010070)
	/usr/local/go/src/runtime/panic.go:481 +0x3e6
github.com/docker/swarmkit/manager/allocator/networkallocator.(*portAllocator).isPortsAllocated(0xc82001e278, 0xc82170b500, 0xc820811880)
	/usr/src/docker/vendor/src/github.com/docker/swarmkit/manager/allocator/networkallocator/portallocator.go:164 +0x4c
github.com/docker/swarmkit/manager/allocator/networkallocator.(*NetworkAllocator).IsServiceAllocated(0xc82075cd50, 0xc82170b500, 0xc82175ac20)
	/usr/src/docker/vendor/src/github.com/docker/swarmkit/manager/allocator/networkallocator/networkallocator.go:309 +0xeb
github.com/docker/swarmkit/manager/allocator.(*Allocator).doNetworkAlloc(0xc82075c2a0, 0x7f861aed9a38, 0xc8201835c0, 0x1b1c4c0, 0xc82175ac20)
	/usr/src/docker/vendor/src/github.com/docker/swarmkit/manager/allocator/network.go:318 +0x369
github.com/docker/swarmkit/manager/allocator.(*Allocator).(github.com/docker/swarmkit/manager/allocator.doNetworkAlloc)-fm(0x7f861aed9a38, 0xc8201835c0, 0x1b1c4c0, 0xc82175ac20)
	/usr/src/docker/vendor/src/github.com/docker/swarmkit/manager/allocator/allocator.go:117 +0x48
github.com/docker/swarmkit/manager/allocator.(*Allocator).run(0xc82075c2a0, 0x7f861aed9a38, 0xc8201835c0, 0xc82075a000, 0xc82075cc00, 0x1d44b28, 0x7, 0xc8208a1680, 0xc8208a1670)
	/usr/src/docker/vendor/src/github.com/docker/swarmkit/manager/allocator/allocator.go:175 +0x118
github.com/docker/swarmkit/manager/allocator.(*Allocator).Run.func2.1(0xc8208a1600, 0xc82075c2a0, 0xc8208a15e0, 0xc82075a000, 0xc82075cc00, 0x1d44b28, 0x7, 0xc8208a1680, 0xc8208a1670)
	/usr/src/docker/vendor/src/github.com/docker/swarmkit/manager/allocator/allocator.go:142 +0xa9
created by github.com/docker/swarmkit/manager/allocator.(*Allocator).Run.func2
	/usr/src/docker/vendor/src/github.com/docker/swarmkit/manager/allocator/allocator.go:143 +0x13c
```

It leads to the restart of engine. In some case, it brings severe consequence.

--- a/manager/allocator/allocator_test.go
+++ b/manager/allocator/allocator_test.go
@@ -50,7 +50,17 @@ func TestAllocator(t *testing.T) {
 						},
 					},
 				},
-				Endpoint: &api.EndpointSpec{},
+				Endpoint: &api.EndpointSpec{
+					Mode: api.ResolutionModeVirtualIP,
+					Ports: []*api.PortConfig{
+						{
+							Name:          "portName",
+							Protocol:      api.ProtocolTCP,
+							TargetPort:    8000,
+							PublishedPort: 8001,
+						},
+					},
+				},
 			},
 		}
 		assert.NoError(t, store.CreateService(tx, s1))
@@ -241,6 +251,16 @@ func TestAllocator(t *testing.T) {
 	}))
 	watchNetwork(t, netWatch, false, isValidNetwork)
 	watchNetwork(t, netWatch, true, nil)
+
+	// Try updating service which is already allocated with no endpointSpec
+	assert.NoError(t, s.Update(func(tx store.Tx) error {
+		s := store.GetService(tx, "testServiceID1")
+		s.Spec.Endpoint = nil
+
+		assert.NoError(t, store.UpdateService(tx, s))
+		return nil
+	}))
+	watchService(t, serviceWatch, false, nil)
 
 	// Try updating task which is already allocated
 	assert.NoError(t, s.Update(func(tx store.Tx) error {

--- a/manager/allocator/network.go
+++ b/manager/allocator/network.go
@@ -564,7 +564,9 @@ func (a *Allocator) allocateNode(ctx context.Context, nc *networkContext, node *
 
 func (a *Allocator) allocateService(ctx context.Context, nc *networkContext, s *api.Service) error {
 	if s.Spec.Endpoint != nil {
+		// service has user-defined endpoint
 		if s.Endpoint == nil {
+			// service currently has no allocated endpoint, need allocated.
 			s.Endpoint = &api.Endpoint{
 				Spec: s.Spec.Endpoint.Copy(),
 			}
@@ -586,6 +588,12 @@ func (a *Allocator) allocateService(ctx context.Context, nc *networkContext, s *
 				s.Endpoint.VirtualIPs = append(s.Endpoint.VirtualIPs,
 					&api.Endpoint_VirtualIP{NetworkID: nc.ingressNetwork.ID})
 			}
+		}
+	} else if s.Endpoint != nil {
+		// service has no user-defined endpoints while has already allocated network resources,
+		// need deallocated.
+		if err := nc.nwkAllocator.ServiceDeallocate(s); err != nil {
+			return err
 		}
 	}
 

--- a/manager/allocator/networkallocator/portallocator.go
+++ b/manager/allocator/networkallocator/portallocator.go
@@ -155,7 +155,18 @@ func (pa *portAllocator) serviceDeallocatePorts(s *api.Service) {
 }
 
 func (pa *portAllocator) isPortsAllocated(s *api.Service) bool {
-	if s.Endpoint == nil {
+	// If service has no user-defined endpoint and allocated endpoint,
+	// we assume it is allocated and return true.
+	if s.Endpoint == nil && s.Spec.Endpoint == nil {
+		return true
+	}
+
+	// If service has allocated endpoint while has no user-defined endpoint,
+	// we assume allocated endpoints are redudant, and they need deallocated.
+	// If service has no allocated endpoint while has user-defined endpoint,
+	// we assume it is not allocated.
+	if (s.Endpoint != nil && s.Spec.Endpoint == nil) ||
+		(s.Endpoint == nil && s.Spec.Endpoint != nil) {
 		return false
 	}
 

